### PR TITLE
Adds support for accessing the Tin Can lesson package file from the web service

### PR DESCRIPTION
**Pull request by mediabounds**
Allows clients to download the package file to serve locally (in case they want to display the course in an offline environment).

### CODE REVIEW

File: lib.php [hunk 1]
```diff
@@ -416,23 +416,20 @@ function tincanlaunch_pluginfile($course, $cm, $context, $filearea, $args, $forc
...
 
+    $filename = array_pop($args);
+    $filepath = implode('/', $args);
     if ($filearea === 'content') {
-        $filename = array_pop($args);
-        $filepath = implode('/', $args);
         $lifetime = null;
     } else if ($filearea === 'package') {
-        $relativepath = implode('/', $args);
-        $fullpath = "/$context->id/tincanlaunch/package/0/$relativepath";
```
Comment (garemoko) [inline lib.php:425]: Why is this deleted? Won't this break uploaded packages?

**Review by garemoko — COMMENTED** [review]
Thanks for this @mediabounds. I added 1 question in the code. 

Looks like the test fails are as a result of tests getting stricter rather than changes introduced in this PR. 

Happy to merge this once it's been tested. In particular I'd like somebody to test that the 'normal' process still works and that files are still securely stored. 

Actually, I guess the functionality of this could be a security concern for content providers. I wonder if it's worth making this a configuration setting to turn it off. What do you think?

**Comment by garemoko** [comment]
@nadavkav @davidpesce @kelvinsleonardo @ryangadams @bell87 @AviMoto or anybody else fancy testing and reviewing this?

File: lib.php (hunk 1 above)
Comment (mediabounds) [reply lib.php:425]: It doesn't _seem_ too--in fact, if you look at `master`, `$fullpath` is never used. This looks to be copied from the SCORM module, but the SCORM module uses this value for calculating the SHA1 to look up the file (which this module does not do).

**Comment by mediabounds** [comment]
In terms of a security concern--I'll defer to someone with more expertise in Moodle. However, the Web service is not on by default, and I believe the ability to download files from the Web service is also off by default.

**Comment by davidpesce** [comment]
I should be able to check it out tomorrow.

**Comment by nadavkav** [comment]
There is a proper way to add web services functionality to Moodle plugins, better migrate your code into Moodle conventional functions and security checks:
https://docs.moodle.org/dev/Adding_a_web_service_to_a_plugin

You can review the following WS functions:
https://docs.moodle.org/dev/Web_service_API_functions
core_course_get_contents
https://github.com/moodle/moodle/blob/master/course/externallib.php#L81

core_files_get_files
https://github.com/moodle/moodle/blob/master/files/externallib.php#L80

And a WS client to test your code:
https://github.com/moodlehq/sample-ws-clients/blob/master/PHP-HTTP-filehandling/client.php

**Comment by mediabounds** [comment]
Thanks, nadavkav--I think this falls in line with those suggestions.

`core_course_get_contents` might return a response that looks like this for a course that contains a single Tin Can Launch Link:

```
[
  {
    "id": 50,
    "name": " 0",
    "visible": 1,
    "summary": "",
    "summaryformat": 1,
    "modules": [
      {
        "id": 39,
        "url": "http://example.com/moodle/mod/tincanlaunch/view.php?id=39",
        "name": "Tin Can Activity",
        "instance": 14,
        "visible": 1,
        "modicon": "http://example.com/moodle/theme/image.php?theme=darkb&component=tincanlaunch&image=icon",
        "modname": "tincanlaunch",
        "modplural": "Tin Can Launch Links",
        "availability": null,
        "indent": 0
      }
    ]
  }
]
```

This works great if your application is always online; you can simply redirect the user to the `url` parameter provided in the response.

However, if you want to allow the user to take the course offline, you need a way to download the course. `course_files_get_files` could be a good way to do this, but it doesn't seem like `course_contents_get_contents` exposes enough information to be able to use that API (namely, the required `itemid` and `filepath` parameters).

This change modifies what `core_course_get_contents` returns with the lesson. So now, it looks like this:

```
[
  {
    "id": 50,
    "name": " 0",
    "visible": 1,
    "summary": "",
    "summaryformat": 1,
    "modules": [
      {
        "id": 39,
        "url": "http://example.com/moodle/mod/tincanlaunch/view.php?id=39",
        "name": "Tin Can Activity",
        "instance": 14,
        "visible": 1,
        "modicon": "http://example.com/moodle/theme/image.php?theme=darkb&component=tincanlaunch&image=icon",
        "modname": "tincanlaunch",
        "modplural": "Tin Can Launch Links",
        "availability": null,
        "indent": 0,
        "contents": [
          {
            "type": "file",
            "filename": "lesson.zip",
            "filepath": "/",
            "filesize": 3201308,
            "fileurl": "http://example.com/moodle/webservice/pluginfile.php/127/mod_tincanlaunch/package/lesson.zip?forcedownload=1",
            "timecreated": 1479243947,
            "timemodified": 1479244330,
            "sortorder": 0,
            "userid": 2,
            "author": "Author",
            "license": "allrightsreserved"
          }
        ]
      }
    ]
  }
]
```

This still respects the authentication mechanisms in place for accessing the API, and `fileurl` still requires an authentication token to download the lesson.

This is similar to how the `mod_resource` plugin works (and is where the idea came from to do it like this).

**Comment by mediabounds** [comment]
In case it helps, here is where this `$modulename_export_contents` hook is called by Moodle:
https://github.com/moodle/moodle/blob/master/course/externallib.php#L284

**Comment by nadavkav** [comment]
Looks great.
You should probably get in contact with @jleyva who is developing Moodle Mobile app, which was responsible for most (if not all) the work related to offline mode. he could give great advice about this.

**Comment by davidpesce** [comment]
I attempted this, but ran into issues (which I believe are) related to my install. I should be able to try again from a fresh install next week.

**Comment by garemoko** [comment]
@davidpesce did you get a chance to re-test this?

**Comment by garemoko** [comment]
This has been open for over a month and has had some testing. Going to merge and test with some other things for the next release.
